tests/test_updater.py

```diff
@@ -1213,6 +1213,73 @@ def test_6_get_one_valid_targetinfo(self):
 
 
 
+  def test_download_target_custom_download_handler(self):
+    # Create temporary directory (destination directory of downloaded targets)
+    # that will be passed as an argument to 'download_target()'.
+    destination_directory = self.make_temp_directory()
+    target_filepaths = \
+      list(self.repository_updater.metadata['current']['targets']['targets'].keys())
+
+    # Get the target info, which is an argument to 'download_target()'.
+    target_filepath = target_filepaths.pop()
+    targetinfo = self.repository_updater.get_one_valid_targetinfo(target_filepath)
+    download_filepath = os.path.join(destination_directory, target_filepath)
+
+    # Test passing a handler with correct signature but incorrect implementation
+    def test_1(param_1, param2):
+      pass
+
+    self.assertRaises(tuf.exceptions.NoWorkingMirrorError,
+        self.repository_updater.download_target, targetinfo, destination_directory,
+        custom_download_handler=test_1)
+
+    # Checks if the file is permanently stored
+    self.assertFalse(os.path.exists(download_filepath))
+
+    # Test passing a handler with incorrect number of parameters
+    def test_2(param_1, param_2, param_3):
+      pass
+
+    with self.assertRaises(tuf.exceptions.NoWorkingMirrorError) as cm:
+      self.repository_updater.download_target(targetinfo, destination_directory,
+          custom_download_handler=test_2)
+
+    for _, mirror_error in six.iteritems(cm.exception.mirror_errors):
+      self.assertTrue(isinstance(mirror_error, TypeError))
+
+    # Checks if the file is permanently stored
+    self.assertFalse(os.path.exists(download_filepath))
+
+    # Test passing a handler throwing an unexpected error
+    def test_3(param_1, param2):
+      raise IOError
+
+    with self.assertRaises(tuf.exceptions.NoWorkingMirrorError) as cm:
+      self.repository_updater.download_target(targetinfo, destination_directory,
+          custom_download_handler=test_3)
+
+    for _, mirror_error in six.iteritems(cm.exception.mirror_errors):
+      self.assertTrue(isinstance(mirror_error, IOError))
+
+    # Checks if the file has been successfully downloaded
+    self.assertFalse(os.path.exists(download_filepath))
+
+    # Test: default case.
+    self.repository_updater.download_target(targetinfo, destination_directory)
+
+    # Checks if the file has been successfully downloaded
+    self.assertTrue(os.path.exists(download_filepath))
+
+    # Test: normal case.
+    self.repository_updater.download_target(targetinfo, destination_directory,
+        custom_download_handler=tuf.download.safe_download)
```

Comment:
I tried! Here it is: ca6f280 
What I did is a very rudimentary implementation of the logic in `tuf.download` (skipping all extra checks, logging, etc).
I am hoping that this will help demonstrating what `tuf` does without the need to follow all the details.
Not sure if this was your expectation?

Reply:
Yeah this is good. I don't think it needs to demonstrate what tuf does: it needs to demonstrate how a custom handler could be written -- and it does that. I have opinions about including SlowRetrievalError but since it relates to the larger issue of what we would require from the handler, I'll write that in a separate comment.